doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,291 @@
+			for {
+
+            // ...omitted
+            
+			}
+		}(); err != nil {
+			panic(err)
+		}
+	}
+```
+
+### Create the message processing loop
+
+If you have trouble following this Go code,
+just read the text
+to get an idea of what you need to do.
+
+First, 
+we read a message
+from our Kafka queue.
+Note the use of a 5-second timeout
+on the read.
+
+```Go
+			// this is the message loop
+            for {
+				// read a message
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				m, err := reader.ReadMessage(ctx)
+				if err != nil {
+					return nil
+                }
+```
+
+Then we create a filename
+and write that to the `tar` stream
+we opened at the beginning
+of the file loop.
+```Go
+                // give it a unique name
+				name := topic + time.Now().Format(time.RFC3339Nano)
+
+                // write the header
+				for err = tw.WriteHeader(&tar.Header{
+					Name: name,
+					Mode: 0600,
+					Size: int64(len(m.Value)),
+				}); err != nil; {
+					if !strings.Contains(err.Error(), "broken pipe") {
+						return err
+					}
+					// if there's a broken pipe, just give it some time to get ready for the next message
+					time.Sleep(5 * time.Millisecond)
+				}
+```
+
+Then we write the actual message 
+as the contents of the file.
+Note the use of a timeout
+in case the named pipe is broken 
+for some reason.
+In this example,
+we're simply moving on to the next message.
+A production system would likely have logging
+and queue operations
+so no messages get missed!
+```Go
+                
+				// and the message
+                              
+				for _, err = tw.Write(m.Value); err != nil; {
+					if !strings.Contains(err.Error(), "broken pipe") {
+						return err
+					}
+					// if there's a broken pipe, just give it some time to get ready for the next message
+					time.Sleep(5 * time.Millisecond)
+				}
+				return nil
+			}
+```
+
+That's the rough outline of operations
+for processing data in queues
+and writing it to a spout
+in a server.
+
+### Create the container for the server
+Once you have containerized your server,
+you can place it in a Pachyderm spout
+by writing an appropriate pipeline specification.
+
+## Writing the spout pipeline specification
+
+A spout is defined using a similar spec
+to a pipeline, 
+and created using the `pachctl create pipeline` command.
+
+In our example above,
+we created a server using Go.
+The Dockerfile for creating a container
+with that server in it is in the [kafka example](https://github.com/pachyderm/pachyderm/tree/master/examples/kafka).
+
+Assuming that's been done,
+we might define the specification 
+for the spout
+something like this.
+
+```json
+{
+  "pipeline": {
+    "name": "my-spout"
+  },
+  "transform": {
+    "cmd": [ "go", "run", "./main.go" ],
+    "image": "myaccount/myimage"
+  },
+  "env": {
+    "HOST": "kafkahost",
+    "TOPIC": "mytopic",
+    "PORT": "9092"
+  },
+  "spout": true
```

Comment:
This needs to be `"spout": {}` -- it's a struct. Even though it's empty, having it specified explicitly will turn this pipeline into a spout because this field has the `json:",omitempty"` tag.

Also, this guide should probably have something about the `spout.overwrite` boolean field. If true, this is equivalent to having the spout use the `--overwrite` flag on all the put files it does. 

Maybe it would be better to just set this explicitly as a best practice, i.e. 
```"spout": {
    "overwrite": false
}``` 
even though this is the same as the `{}` version.

Reply:
I like establishing this practice, and the extra info here.  Thank you!